File: scripts/test-voice-call-diagnostics.js
const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const source = fs.readFileSync(path.join(__dirname, '..', 'web', 'src', 'voice-call-diagnostics.js'), 'utf8');

function makeStorage(seed = {}) {
  const values = new Map(Object.entries(seed));
  return {
    getItem(key) { return values.has(key) ? values.get(key) : null; },
    setItem(key, value) { values.set(key, String(value)); },
    removeItem(key) { values.delete(key); },
    dump() { return Object.fromEntries(values); },
  };
}

function load(storage, reports) {
  const window = {
    crypto: { randomUUID: () => '11111111-2222-4333-8444-555555555555' },
    localStorage: storage,
    location: { href: 'https://app.munea.net/index.html' },
    console: { info() {} },
  };
  const context = vm.createContext({ window, URL, Date, Math, JSON, String, Number, Array, Object, RegExp });
  vm.runInContext(source, context, { filename: 'voice-call-diagnostics.js' });
  if (reports) window.MuneaVoiceDiagnostics.setReporter((eventName, properties) => reports.push({ eventName, properties }));
  return window.MuneaVoiceDiagnostics;
}

const storage = makeStorage();
const reports = [];
const diag = load(storage, reports);
diag.start({
  routeMode: 'gateway',
  voiceEndpoint: 'wss://voice.example/ws?token=do-not-store',
  avatarEndpoint: 'https://avatar.example/offer?key=do-not-store',
  transcript: 'private words',
});
diag.mark('gateway_assigned', 'pass', { gatewayCallId: 'call-123', token: 'secret-token' });
diag.mark('voice_socket_open', 'pass', { endpoint: 'wss://voice.example/ws?key=secret' });
diag.fail('voice_socket_closed', 'ws_4403', { closeCode: 4403, closeReason: 'call token required' });
const summary = diag.end('failed', { reason: 'voice_socket_closed' });

assert.strictEqual(summary.outcome, 'failed');
assert.strictEqual(summary.lastSuccessfulStage, 'voice_socket_open');
assert.strictEqual(summary.firstFailedStage, 'voice_socket_closed');
assert(summary.stages.some(stage => stage.stage === 'gateway_assigned'));
const serialized = JSON.stringify(storage.dump());
assert(!serialized.includes('do-not-store'));
assert(!serialized.includes('secret-token'));
assert(!serialized.includes('private words'));
assert(reports.some(item => item.eventName === 'voice_call_stage_failed'));
const terminal = reports.find(item => item.eventName === 'voice_call_diagnostic');
assert(terminal, 'terminal diagnostic summary must be reported');
assert(terminal.properties.stages.some(stage => stage.stage === 'voice_socket_open'));

diag.start({ appVersion: '1.0.18' });
const cancelled = diag.end('cancelled', { reason: 'user_cancelled' });
assert.strictEqual(cancelled.outcome, 'cancelled');

const abandonedStorage = makeStorage();
const firstLoad = load(abandonedStorage);
firstLoad.start({ routeMode: 'development_direct' });
firstLoad.mark('voice_socket_open', 'pass');
const recoveredReports = [];
load(abandonedStorage, recoveredReports);
const abandoned = recoveredReports.find(item => item.eventName === 'voice_call_diagnostic');
assert(abandoned, 'unfinished trace must be reported after the next launch');
assert.strictEqual(abandoned.properties.outcome, 'abandoned');
assert.strictEqual(abandoned.properties.reason, 'app_terminated_or_reloaded');

const appSource = fs.readFileSync(path.join(__dirname, '..', 'web', 'src', 'app.js'), 'utf8');
const indexSource = fs.readFileSync(path.join(__dirname, '..', 'web', 'index.html'), 'utf8');
assert(indexSource.indexOf('src/voice-call-diagnostics.js') < indexSource.indexOf('src/app.js'), 'diagnostics must load before the App module');
[
  'microphone_ready',
  'gateway_assigned',
  'voice_socket_open',
  'voice_ready',
  'avatar_offer_accepted',
  'avatar_first_frame',
  'opening_audio_ready',
  'voice_first_audio',
  'asr_first_caption',
  'call_connected',
].forEach(stage => assert(appSource.includes("'" + stage + "'"), 'missing App trace stage: ' + stage));
assert(source.includes("mark('dial_tapped'"), 'trace must start at the call button tap');
assert(appSource.includes('closeCode: event.code') && appSource.includes('closeReason: event.reason'), 'WebSocket close code and reason must be captured');

console.log('Voice call diagnostics contracts: PASS');

File: engine/test_voice_call_diagnostics.py
#!/usr/bin/env python3

import os
import sys
from unittest import mock

os.environ.setdefault("GEMINI_API_KEY", "voice-diagnostics-test-key")
os.environ.setdefault("MUNEA_DATABASE_PROVIDER", "json")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import server


EVENTS = [
    {
        "eventName": "voice_call_diagnostic",
        "eventTime": "2026-07-15T10:02:00Z",
        "properties": {
            "callId": "call-failed",
            "outcome": "failed",
            "reason": "readiness_timeout",
            "firstFailedStage": "avatar_first_frame",
            "lastSuccessfulStage": "voice_ready",
            "totalMs": 30000,
            "context": {
                "appVersion": "1.0.18",
                "routeMode": "gateway",
                "voiceEndpoint": "wss://voice.example/",
                "avatarEndpoint": "https://avatar.example/",
            },
        },
    },
    {
        "eventName": "voice_call_diagnostic",
        "eventTime": "2026-07-15T10:01:00Z",
        "properties": {
            "callId": "call-ok",
            "outcome": "completed",
            "reason": "user_ended",
            "lastSuccessfulStage": "call_connected",
            "totalMs": 90000,
            "context": {"appVersion": "1.0.18", "routeMode": "development_direct"},
        },
    },
    {
        "eventName": "voice_call_diagnostic",
        "eventTime": "2026-07-15T10:00:00Z",
        "properties": {
            "callId": "call-malformed",
            "outcome": "failed",
            "reason": "invalid_client_payload",
            "firstFailedStage": "voice_socket_connecting",
            "totalMs": "not-a-number",
            "context": {
                "voiceEndpoint": "wss://user:super-secret@voice.example/connect?token=hidden-token",
                "avatarEndpoint": "https://avatar.example/offer?key=hidden-key#debug",
            },
        },
    },
    {"eventName": "voice_session_started", "properties": {}},
]


def main():
    with mock.patch.object(server, "load_product_events", return_value=EVENTS), \
            mock.patch.object(server, "data_backend_status", return_value={"enabled": True}):
        result = server.admin_voice_diagnostics_summary({"days": 7, "limit": 10})

    assert result["ok"] is True
    assert result["count"] == 3
    assert result["successRate"] == 0.3333
    assert result["totals"]["byOutcome"] == {"completed": 1, "failed": 2}
    assert result["totals"]["byFailedStage"] == {
        "avatar_first_frame": 1,
        "voice_socket_connecting": 1,
    }
    assert result["totals"]["averageTotalMs"] == 40000
    assert result["recent"][0]["callId"] == "call-failed"
    assert result["recent"][0]["lastSuccessfulStage"] == "voice_ready"
    assert result["privacy"] == {
        "rawAudioStored": False,
        "rawTranscriptStored": False,
        "credentialsStored": False,
    }
    serialized = str(result)
    assert "stages" not in result["recent"][0]
    assert "private words" not in serialized.lower()
    assert "super-secret" not in serialized
    assert "hidden-token" not in serialized
    assert "hidden-key" not in serialized
    malformed = next(item for item in result["recent"] if item["callId"] == "call-malformed")
    assert malformed["totalMs"] == 0
    assert malformed["voiceEndpoint"] == "wss://voice.example/connect"
    assert malformed["avatarEndpoint"] == "https://avatar.example/offer"
    print("Voice call diagnostics admin summary: PASS")


if __name__ == "__main__":
    main()
